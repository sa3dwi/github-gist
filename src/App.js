import React, { useState, useEffect, useCallback, useMemo } from 'react';

import { appConfig } from "./config/app.config";
import searchAction from "./actions/searchAction";
import useLocalStorage from "./hooks/useLocalStorage";

import SearchList from './components/search/SearchList';

const App = () => {
  const [searchQuery, setSearchQuery] = useState('');
  const [submitAction, setSubmitAction] = useState(false);
  const [searchResults, setSearchResults] = useState([]);
  const [errorMessage, setErrorMessage] = useState('No Results found');

  // use custom local storage hook to save resent search queries 
  const [resentSearch, setResentSearch, clearResentSearch] = useLocalStorage(appConfig.LOCAL_STORAGE_KEY,'');

  // input form handles 
  const handleChange = text => setSearchQuery(text);
  const handleClearResent = e => {
    e.preventDefault();
    clearResentSearch()
  };
  const handleSubmit = () => {
      setSubmitAction(true)
  };

  // submit when onKeyPress enter 
  const handleKeyPress = e => {
    if (e.key === 'Enter' && searchQuery.length > appConfig.MIN_SEARCH_CHARACTERS)
      setSubmitAction(true)
  }

  // search action
  const searchAct = useCallback(searchQuery => searchAction.getGistByUserName(searchQuery),[]);

  useEffect(() => {
    // check min search characters
    if(submitAction && searchQuery.length > appConfig.MIN_SEARCH_CHARACTERS) {
      // show loding msg 
      setErrorMessage('Loading...');
      // call search action with query 
      searchAct(searchQuery).then((res) => {
        // set serach results or show error msg
        if (res.message) {
          setErrorMessage(res.message);
        } else {
          setSearchResults(res);
          setErrorMessage(false);
        }
      }).finally( () => {
        // clear the side effects 
        setResentSearch(searchQuery);
        setSubmitAction(false);
      });
    }
  }, [submitAction, searchQuery, searchAct, setResentSearch])

  // memoizing search List Component
  const searchListComponant = useMemo(() => {
    return (<SearchList searchResults={searchResults} errorMessage={errorMessage} />)
  }, [searchResults, errorMessage])

  return (
    <>
      <header>
          <div className="container">
              <div className="MainLogo">
                <h3>
                  <img alt="gitHub_logo" width="64" height="64" src="images/gitHub_logo.png"/>
                  GitHub Gist Search
                </h3>
                <h6>Search with username below</h6>
              </div>
              <div className="mainSearch">
                      <input type="search" className="MainInput" name="q" id="q"
                      placeholder="Start typing to search .." 
                      autoComplete="off"
                      onChange={e => handleChange(e.target.value)}
                      onKeyPress={e => handleKeyPress(e)}
                      maxLength={appConfig.MAX_SEARCH_CHARACTERS}
                      data-testid="searchInput"/>
                      <button type="submit" data-testid="submitSearchBtn" className="MainInput" onClick={handleSubmit}>Search</button>
              </div>
              <div data-testid="resentSearch">
                {typeof(resentSearch) !== 'undefined' && resentSearch.length > 0 && 
                  <p>
                    <span className="strong7">ResentSearch:</span> {resentSearch.join(',')} <a href="/" className="clear_link" onClick={e => handleClearResent(e)}>Clear resent</a>
                  </p>
                }
              </div>
          </div>
      </header>

      {searchListComponant}
    </>
  );
}

export default App;
