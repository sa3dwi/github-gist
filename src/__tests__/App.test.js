import React from 'react';
import { render, screen, fireEvent, cleanup, waitForElement } from '@testing-library/react'
import { create } from "react-test-renderer";
import { shallow } from 'enzyme';
import { expect } from 'chai';

import App from './../App';

beforeEach(() => {
  localStorage.clear();
  cleanup
})

describe("App component UI", () => {
  test('has correct welcome text', () => {
    const wrapper = shallow(<App />)
    expect(wrapper.find('h3').text()).to.equal('GitHub Gist Search')
  })
  
  test('has button with search text', () => {
    const wrapper = shallow(<App />)
    expect(wrapper.find('button').text()).to.equal('Search')
  })

  test('matches the snapshot', () => {
    const wrapper = create(<App />)
    expect(wrapper.toJSON()).to.matchSnapshot()
  })
})

describe("App component Functionality", () => {
  test('allows you submit search', () => {
    render(<App />)
    const input = screen.getByTestId('searchInput');
    const submit = screen.getByTestId('submitSearchBtn');

    input.value = 'sa3dwi'
    fireEvent.click(submit)
  })

  test('allows you add to resentSearch', () => {
    render(<App />)
    const input = screen.getByTestId('searchInput');
    const submit = screen.getByTestId('submitSearchBtn');
    const resentSearchSection = screen.getByTestId('resentSearch');

    input.value = 'sa3dwi'
    fireEvent.click(submit)

    expect(resentSearchSection).not.to.equal(null);
  })

  test('allows you show empty results msg', () => {
    render(<App />)
    const input = screen.getByTestId('searchInput');
    const submit = screen.getByTestId('submitSearchBtn');
    const searchListWarp = screen.getByTestId('searchListWarp');

    input.value = 'sa3'
    fireEvent.click(submit)

    expect(searchListWarp.textContent).to.equal('No Results found')
  })
})
