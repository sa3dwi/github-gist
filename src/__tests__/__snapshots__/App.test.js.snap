// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component UI matches the snapshot 1`] = `
Array [
  <header>
    <div
      className="container"
    >
      <div
        className="MainLogo"
      >
        <h3>
          <img
            alt="gitHub_logo"
            height="64"
            src="images/gitHub_logo.png"
            width="64"
          />
          GitHub Gist Search
        </h3>
        <h6>
          Search with username below
        </h6>
      </div>
      <div
        className="mainSearch"
      >
        <input
          autoComplete="off"
          className="MainInput"
          data-testid="searchInput"
          id="q"
          maxLength={50}
          name="q"
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="Start typing to search .."
          type="search"
        />
        <button
          className="MainInput"
          data-testid="submitSearchBtn"
          onClick={[Function]}
          type="submit"
        >
          Search
        </button>
      </div>
      <div
        data-testid="resentSearch"
      />
    </div>
  </header>,
  <div
    className="wrap"
    data-testid="searchListWarp"
  >
    <div
      className="container"
    >
      <div
        className="row d-flex justify-content-center alert"
      >
        No Results found
      </div>
    </div>
  </div>,
]
`;
